pkg/kine/prepared/db.go

```diff
@@ -0,0 +1,191 @@
+package prepared
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+type DB struct {
+	underlying *sql.DB
+	mu         sync.Mutex
+	maxSize    int
+	store      map[string]*lruEntry
+	lruList    lruList
+}
+
+func New(db *sql.DB, maxSize int) *DB {
+	return &DB{
+		underlying: db,
+		maxSize:    maxSize,
+		store:      make(map[string]*lruEntry, maxSize),
+	}
+}
+
+func (db *DB) Underlying() *sql.DB { return db.underlying }
+
+func (db *DB) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.ExecContext(ctx, args...)
+}
+
+func (db *DB) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.QueryContext(ctx, args...)
+}
+
+func (db *DB) Close() error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	errs := []error{}
+	for len(db.store) != 0 {
+		if err := db.drop(db.lruList.tail()); err != nil {
```

Comment:
The conditions here seem a little muddled in my mind---we're iterating whilst the store is non-empty, getting the tail of the list each time, which is implicitly changed by calling `db.drop`. Do we need to iterate over the cached queries in a particular order? I think it would be cleaner to do
```go
for _, cacheEntry := range db.store {
    ...
}
```
If the order matters and we do indeed wish to drop the most used one last, I think could be nice to have an iterator over the LRU list like this:
```go
func (ll *lruList) IterateFromLeast() *lruFromLeastIterator {
    return &lruIterator{
        curr: ll.tail()
    }
}

struct lruFromLeastIterator {
    curr *lruEntry
}

func (it *lruFromLeastIterator) Next(entry **lruEntry) bool {
    if it.curr == nil {
        return false
    }
    *entry = it.curr
    it.curr = it.curr.prev
    return true
}

iter := db.lruList.IterateFromLeast()
var entry *lruEntry
for iter.Next(&entry) {
    if err := db.drop(entry); err != nil {
        errs = append(errs, err)
    }
}

Reply:
Removed.